src/Core/ServerSettings.h

```diff
@@ -13,98 +13,119 @@ class AbstractConfiguration;
-    M(UInt64, max_backups_io_thread_pool_free_size, 0, "Max free size for backups IO thread pool.", 0) \
-    M(UInt64, backups_io_thread_pool_queue_size, 0, "Queue size for backups IO thread pool.", 0) \
-    M(UInt64, backup_threads, 16, "The maximum number of threads to execute BACKUP requests.", 0) \
-    M(UInt64, max_backup_bandwidth_for_server, 0, "The maximum read speed in bytes per second for all backups on server. Zero means unlimited.", 0) \
-    M(UInt64, restore_threads, 16, "The maximum number of threads to execute RESTORE requests.", 0) \
-    M(Int32, max_connections, 1024, "Max server connections.", 0) \
-    M(UInt32, asynchronous_metrics_update_period_s, 1, "Period in seconds for updating asynchronous metrics.", 0) \
-    M(UInt32, asynchronous_heavy_metrics_update_period_s, 120, "Period in seconds for updating heavy asynchronous metrics.", 0) \
-    M(String, default_database, "default", "Default database name.", 0) \
-    M(String, tmp_policy, "", "Policy for storage with temporary data.", 0) \
-    M(UInt64, max_temporary_data_on_disk_size, 0, "The maximum amount of storage that could be used for external aggregation, joins or sorting., ", 0) \
-    M(String, temporary_data_in_cache, "", "Cache disk name for temporary data.", 0) \
-    M(UInt64, aggregate_function_group_array_max_element_size, 0xFFFFFF, "Max array element size in bytes for groupArray function. This limit is checked at serialization and help to avoid large state size.", 0) \
-    M(UInt64, max_server_memory_usage, 0, "Maximum total memory usage of the server in bytes. Zero means unlimited.", 0) \
-    M(Double, max_server_memory_usage_to_ram_ratio, 0.9, "Same as max_server_memory_usage but in to RAM ratio. Allows to lower max memory on low-memory systems.", 0) \
-    M(UInt64, merges_mutations_memory_usage_soft_limit, 0, "Maximum total memory usage for merges and mutations in bytes. Zero means unlimited.", 0) \
-    M(Double, merges_mutations_memory_usage_to_ram_ratio, 0.5, "Same as merges_mutations_memory_usage_soft_limit but in to RAM ratio. Allows to lower memory limit on low-memory systems.", 0) \
-    M(Bool, allow_use_jemalloc_memory, true, "Allows to use jemalloc memory.", 0) \
-    M(UInt64, async_insert_threads, 16, "Maximum number of threads to actually parse and insert data in background. Zero means asynchronous mode is disabled", 0) \
-    M(Bool, async_insert_queue_flush_on_shutdown, true, "If true queue of asynchronous inserts is flushed on graceful shutdown", 0) \
+enum class RuntimeReloadType
+{
+    FULL,
+    ONLY_INCREASE,
+    NO,
+};
+
+#define SET_RUNTIME_RELOAD_(M, TYPE, NAME, DEFAULT, DESCRIPTION, FLAGS, RUNTIME_RELOAD) \
+    this->runtime_reload_map.insert(std::make_pair(#NAME, RUNTIME_RELOAD));
+
+#define M_WITH_FLAG_(M, TYPE, NAME, DEFAULT, DESCRIPTION, FLAGS, RUNTIME_RELOAD) \
+    M(TYPE, NAME, DEFAULT, DESCRIPTION, FLAGS) \
```

Comment:
Looks like too many letters, Can we leave just `M` and change a few other macros?

Reply:
I was thinking it should be possible to use M(TYPE, NAME, DEFAULT, DESCRIPTION, FLAGS, RUNTIME_RELOAD) directly, but now I see that `SERVER_SETTINGS` is passed directly to `DECLARE_SETTINGS_TRAITS` which is used in a lot of places, so changing it is not a good idea.

But, yes macro's name still feels too long to me, just because almost everywhere we use M(X, Y, Z) do declare such a list. Let's ask @rschu1ze for his opinion. I was just glad to see this PR :-)

Reply:
@serxa Ok, The new macro name does sound awkward, and I will make sure to modify it later.